core/controllers/retrieve.py

```diff
@@ -0,0 +1,42 @@
+from flask import abort, make_response, request
+from sqlalchemy import select
+
+from core.const import EXCEL_MIMETYPE
+from core.db import db
+from core.db.entities import Programme, ProgrammeProgress, Submission
+
+
+def retrieve():
+    """Handle the download request and return the originally submitted spreadsheet.
+
+    Select file by:
+    - reporting round
+    - programme name
+
+    :return: Flask response object containing the requested spreadsheet.
+    """
+    reporting_round = request.args.get("reporting_round")
+    programme_name = request.args.get("programme_name")
+
+    query = (
+        select(Submission.submission_file)
+        .join(ProgrammeProgress, Submission.id == ProgrammeProgress.submission_id)
+        .join(Programme, ProgrammeProgress.programme_id == Programme.id)
+        .where(Programme.programme_name == programme_name)
+        .where(Submission.reporting_round == reporting_round)
+    )
+
+    file_bytes = db.session.scalars(query).first()
+
+    if not file_bytes:
+        return abort(400, "No file was found that matches these criteria")
+
+    response = make_response(file_bytes)
+    response.headers.set("Content-Type", EXCEL_MIMETYPE)
+    response.headers.set(
+        "Content-Disposition",
+        "attachment",
+        filename=f"{'_'.join(programme_name.split())}_round_{reporting_round}_submission.xlsx",
+    )
+
+    return response
```

Comment:
[flask.send_file](https://tedboy.github.io/flask/generated/flask.send_file.html) may be a cleaner way to do this